src/history/test/HistoryTests.cpp

```diff
@@ -224,91 +225,136 @@ TEST_CASE("Ledger chain verification", "[ledgerheaderverification]")
 
     auto checkExpectedBehavior = [&](Work::State expectedState,
                                      LedgerHeaderHistoryEntry lcl,
-                                     LedgerHeaderHistoryEntry last) {
+                                     LedgerHeaderHistoryEntry last,
+                                     bool hasTrustedHash = true) {
         auto lclPair = LedgerNumHashPair(lcl.header.ledgerSeq,
                                          std::make_optional<Hash>(lcl.hash));
         auto ledgerRangeEnd = LedgerNumHashPair(
-            last.header.ledgerSeq, std::make_optional<Hash>(last.hash));
+            last.header.ledgerSeq, hasTrustedHash
+                                       ? std::make_optional<Hash>(last.hash)
+                                       : std::nullopt);
         std::promise<LedgerNumHashPair> ledgerRangeEndPromise;
         std::shared_future<LedgerNumHashPair> ledgerRangeEndFuture =
             ledgerRangeEndPromise.get_future().share();
         ledgerRangeEndPromise.set_value(ledgerRangeEnd);
+
+        auto fataFailurePromise = std::promise<bool>();
+        std::shared_future<bool> fatalFailureFuture =
+            fataFailurePromise.get_future().share();
         auto w = wm.executeWork<VerifyLedgerChainWork>(
-            tmpDir, ledgerRange, lclPair, ledgerRangeEndFuture);
+            tmpDir, ledgerRange, lclPair, ledgerRangeEndFuture,
+            std::move(fataFailurePromise));
         REQUIRE(expectedState == w->getState());
+        REQUIRE(fatalFailureFuture.valid());
+        return fatalFailureFuture.get();
     };
 
     LedgerHeaderHistoryEntry lcl, last;
-    LOG_DEBUG(DEFAULT_LOG, "fully valid");
+    SECTION("fully valid")
```

Comment:
Just checking -- the change to `SECTION`s makes the test re-run once for each `SECTION` -- is this intentional? eg. to reset the fatal failure promise?

Reply:
This is intentional: this test was broken before. Every "section" generated ledger chain files, which I think all shared the same directory/files. Out XDR output stream opens files in append mode, so all sections below "fully valid" were actually failing with hash mismatch (since we pass a new trusted hash, but verify-ledger-chain would read old data from "fully valid" section). Of course, we haven't noticed this for years, because the test only checks that the resulting work is in WORK_FAILURE state :/

Reply:
oh wow. I'm .. very sad to hear this indeed. thanks for fixing.

Reply:
oh wow. I'm .. very sad to hear this indeed. thanks for fixing.